wpe/src/main/cpp/Browser/CookieManager.cpp

```diff
@@ -0,0 +1,117 @@
+ */
+
+#include "CookieManager.h"
+
+#include <wpe/webkit.h>
+
+#include "Browser.h"
+#include "JNI/JNI.h"
+#include "Logging.h"
+
+DECLARE_JNI_CLASS_SIGNATURE(JNICallback, "com/wpe/wpeview/WPECallback");
+DECLARE_JNI_CLASS_SIGNATURE(JNICookieManager, "com/wpe/wpeview/WPECookieManager");
+
+class JNICallbackCache final : public JNI::TypedClass<JNICallback> {
+public:
+    JNICallbackCache()
+        : JNI::TypedClass<JNICallback>(true)
+        , m_onBooleanResult(getStaticMethod<void(JNICallback, jboolean)>("onBooleanResult"))
+    {
+    }
+
+    void onBooleanResult(JNICallback callback, jboolean result) const noexcept
+    {
+        try {
+            m_onBooleanResult.invoke(callback, result);
+        } catch (const std::exception& ex) {
+            Logging::logError("cannot call WPECookieManager callback result (%s)", ex.what());
+        }
+        JNIEnv* env = JNI::getCurrentThreadJNIEnv();
+        env->DeleteGlobalRef(callback);
+    }
+
+private:
+    const JNI::StaticMethod<void(JNICallback, jboolean)> m_onBooleanResult;
+};
+
+const JNICallbackCache& getJNICallbackCache()
+{
+    static const JNICallbackCache s_singleton;
+    return s_singleton;
+}
+
+class JNICookieManagerCache final : public JNI::TypedClass<JNICookieManager> {
+public:
+    JNICookieManagerCache();
+
+    static void onRemoveAllCookiesReady(WebKitWebsiteDataManager* manager, GAsyncResult* result, JNICallback callback)
+    {
+        gboolean clearResult = webkit_website_data_manager_clear_finish(manager, result, nullptr);
+        getJNICallbackCache().onBooleanResult(callback, static_cast<jboolean>(clearResult));
+    }
+
+private:
+    static void nativeSetCookieAcceptPolicy(JNIEnv* env, jclass cls, jint jpolicy);
+    static void nativeRemoveAllCookies(JNIEnv* env, jclass cls, JNICallback callback);
+};
+
+const JNICookieManagerCache& getJNICookieManagerCache()
+{
+    static const JNICookieManagerCache s_singleton;
+    return s_singleton;
+}
+
+JNICookieManagerCache::JNICookieManagerCache()
+    : JNI::TypedClass<JNICookieManager>(true)
+{
+    registerNativeMethods(JNI::StaticNativeMethod<void(jint)>(
+                              "nativeSetCookieAcceptPolicy", JNICookieManagerCache::nativeSetCookieAcceptPolicy),
+        JNI::StaticNativeMethod<void(JNICallback)>(
+            "nativeRemoveAllCookies", JNICookieManagerCache::nativeRemoveAllCookies));
+}
+
+void JNICookieManagerCache::nativeSetCookieAcceptPolicy(JNIEnv* /*env*/, jclass /*cls*/, jint jpolicy)
+{
+    auto toWebKitPolicy = [](CookieManager::CookieAcceptPolicy policy) -> WebKitCookieAcceptPolicy {
+        switch (policy) {
+        case CookieManager::CookieAcceptPolicy::AcceptNever:
+            return WEBKIT_COOKIE_POLICY_ACCEPT_NEVER;
+        case CookieManager::CookieAcceptPolicy::AcceptNoThirdParty:
+            return WEBKIT_COOKIE_POLICY_ACCEPT_NO_THIRD_PARTY;
+        default:
+            return WEBKIT_COOKIE_POLICY_ACCEPT_ALWAYS;
+        }
+    };
```

Comment:
I thought about that also but then I started thinking that they might diverge in the future. For example if webkit for some reason changes their enum values for accept policy in future releases. Then it would be hard to catch bug. Thus this fail safe approach

Reply:
I understand your point but if values change in wpewebkit (which would be quite surprising for this enum), in all cases you will have to modify this code.

I think it will be a lot easier to just change the values in the Java code. Indeed the name of the enum in the Java code is: `NativeCookieAcceptPolicy`. It is explicit that it should contains the native values for this enum.

Here this indirection adds confusion and compel the developer to go to different places to update something that would be already straightforward with just the  `NativeCookieAcceptPolicy` enum in the Java code.